Explorer/Assets/DCL/VoiceChat/VoiceChatMicrophoneHandler.cs

```diff
@@ -0,0 +1,375 @@
+using DCL.Diagnostics;
+using DCL.Settings.Settings;
+using System;
+using UnityEngine;
+using UnityEngine.InputSystem;
+using Cysharp.Threading.Tasks;
+using System.Threading;
+using Utility;
+using Object = UnityEngine.Object;
+
+namespace DCL.VoiceChat
+{
+    public class VoiceChatMicrophoneHandler : IDisposable
+    {
+        private const bool MICROPHONE_LOOP = true;
+        private const int MICROPHONE_LENGTH_SECONDS = 1;
+
+        private readonly VoiceChatSettingsAsset voiceChatSettings;
+        private readonly VoiceChatConfiguration voiceChatConfiguration;
+        private readonly VoiceChatMicrophoneAudioFilter audioFilter;
+        private readonly AudioSource audioSource;
+
+        private AudioClip microphoneAudioClip;
+        private bool isMicrophoneInitialized;
+        private bool isInCall;
+        private bool isTalking { get; set; }
+        private CancellationTokenSource microphoneChangeCts;
+        private int microphoneSampleRate;
+        private string microphoneName;
+        private float buttonPressStartTime;
+
+
+        public VoiceChatMicrophoneAudioFilter AudioFilter => audioFilter;
+        public event Action EnabledMicrophone;
+        public event Action DisabledMicrophone;
+
+        public VoiceChatMicrophoneHandler(
+            VoiceChatSettingsAsset voiceChatSettings,
+            VoiceChatConfiguration voiceChatConfiguration,
+            AudioSource audioSource,
+            VoiceChatMicrophoneAudioFilter audioFilter)
+        {
+            this.voiceChatSettings = voiceChatSettings;
+            this.voiceChatConfiguration = voiceChatConfiguration;
+            this.audioSource = audioSource;
+            this.audioFilter = audioFilter;
+
+            DCLInput.Instance.VoiceChat.Talk.performed += OnPressed;
+            DCLInput.Instance.VoiceChat.Talk.canceled += OnReleased;
+            voiceChatSettings.MicrophoneChanged += OnMicrophoneChanged;
+            isInCall = false;
+        }
+
+        public void Dispose()
+        {
+            DCLInput.Instance.VoiceChat.Talk.performed -= OnPressed;
+            DCLInput.Instance.VoiceChat.Talk.canceled -= OnReleased;
+            voiceChatSettings.MicrophoneChanged -= OnMicrophoneChanged;
+
+            microphoneChangeCts?.SafeCancelAndDispose();
+
+            if (isMicrophoneInitialized)
+            {
+                if (audioSource != null)
+                {
+                    audioSource.volume = 0f;
+                    StopAudioSource();
+                    audioSource.clip = null;
+                }
+
+                Microphone.End(microphoneName);
+
+                if (audioFilter != null)
+                    audioFilter.enabled = false;
+            }
+        }
+
+        private void StopAudioSource()
+        {
+            if (!PlayerLoopHelper.IsMainThread)
```

Comment:
It's not a big harm if you simply call `StopAudioSourceAsync` from the main thread - it will execute synchronously. But you will erase a bit of the repetitive code